libdnf/repo/repo_downloader.cpp

```diff
@@ -159,172 +123,8 @@ static void move_recursive(const std::filesystem::path & src, const std::filesys
-        handle_set_opt(h, LRO_IPRESOLVE, LR_IPRESOLVE_V4);
-    } else if (ip_resolve == "ipv6") {
-        handle_set_opt(h, LRO_IPRESOLVE, LR_IPRESOLVE_V6);
-    }
-
-    auto userpwd = config.username().get_value();
-    if (!userpwd.empty()) {
-        // TODO Use URL encoded form, needs support in librepo
-        userpwd = format_user_pass_string(userpwd, config.password().get_value(), false);
-        handle_set_opt(h, LRO_USERPWD, userpwd.c_str());
-    }
-
-    if (!config.sslcacert().get_value().empty()) {
-        handle_set_opt(h, LRO_SSLCACERT, config.sslcacert().get_value().c_str());
-    }
-    if (!config.sslclientcert().get_value().empty()) {
-        handle_set_opt(h, LRO_SSLCLIENTCERT, config.sslclientcert().get_value().c_str());
-    }
-    if (!config.sslclientkey().get_value().empty()) {
-        handle_set_opt(h, LRO_SSLCLIENTKEY, config.sslclientkey().get_value().c_str());
-    }
-    long sslverify = config.sslverify().get_value() ? 1L : 0L;
-    handle_set_opt(h, LRO_SSLVERIFYHOST, sslverify);
-    handle_set_opt(h, LRO_SSLVERIFYPEER, sslverify);
-
-    // === proxy setup ===
-    if (!config.proxy().empty() && !config.proxy().get_value().empty()) {
-        handle_set_opt(h, LRO_PROXY, config.proxy().get_value().c_str());
-    }
-
-    const std::string proxy_auth_method_str = config.proxy_auth_method().get_value();
-    auto proxy_auth_method = LR_AUTH_ANY;
-    for (auto & auth : PROXYAUTHMETHODS) {
-        if (proxy_auth_method_str == auth.name) {
-            proxy_auth_method = auth.code;
-            break;
-        }
-    }
-    handle_set_opt(h, LRO_PROXYAUTHMETHODS, static_cast<long>(proxy_auth_method));
-
-    if (!config.proxy_username().empty()) {
-        auto userpwd = config.proxy_username().get_value();
-        if (!userpwd.empty()) {
-            userpwd = format_user_pass_string(userpwd, config.proxy_password().get_value(), true);
-            handle_set_opt(h, LRO_PROXYUSERPWD, userpwd.c_str());
-        }
-    }
-
-    if (!config.proxy_sslcacert().get_value().empty()) {
-        handle_set_opt(h, LRO_PROXY_SSLCACERT, config.proxy_sslcacert().get_value().c_str());
-    }
-    if (!config.proxy_sslclientcert().get_value().empty()) {
-        handle_set_opt(h, LRO_PROXY_SSLCLIENTCERT, config.proxy_sslclientcert().get_value().c_str());
-    }
-    if (!config.proxy_sslclientkey().get_value().empty()) {
-        handle_set_opt(h, LRO_PROXY_SSLCLIENTKEY, config.proxy_sslclientkey().get_value().c_str());
-    }
-    long proxy_sslverify = config.proxy_sslverify().get_value() ? 1L : 0L;
-    handle_set_opt(h, LRO_PROXY_SSLVERIFYHOST, proxy_sslverify);
-    handle_set_opt(h, LRO_PROXY_SSLVERIFYPEER, proxy_sslverify);
-
-    return handle;
-}
-
-
-namespace libdnf::repo {
-
 LibrepoError::LibrepoError(std::unique_ptr<GError> && lr_error)
-    : Error(M_("Librepo error: {}"), lr_error->message),
+    : Error(M_("Librepo error: {}"), std::string(lr_error->message)),
       code(lr_error->code) {}
```

Comment:
It seems this should go into the new `librepo.cpp` file. Also why the added conversion to `std::string`?

Reply:
Error constructor captures (copies) its arguments. But in that case, `lr_error->message` is a C string - a pointer to a dynamicaly allocated null terminated array of characters. Without conversion to `std::string`, the Error constructor will copy the pointer. But the pointed memory is then freed and the copied pointer would point to the freed memory.